BR-4382: (SI 76716) Diagnostics not writing custom_directory.zip file…

alex-vlasov opened the pull request:
… in Shadow

morozov commented on shadow.c:
```diff
@@ -117,10 +117,19 @@ static PHP_GSHUTDOWN_FUNCTION(shadow)
+/* {{{ shadow_deps
+ */
+static const zend_module_dep shadow_deps[] = {
+	ZEND_MOD_OPTIONAL("zip")
```
I wouldn't mention any particular extensions here, since it would solve the problem for the mentioned ones and create a false feeling of the problem having been fixed, but for the rest of extensions it would still exist.

morozov commented:
As an idea, we could try to override the PHP internal function which registers new extensions and throw a warning from there if any extension is being registered after shadow. However, even in this case, we may have to maintain some list of exceptions, since some extensions may rely on being loaded the last.

alex-vlasov commented:
I've added startup code that checks other extensions to be loaded before

morozov commented on shadow.c:
```diff
@@ -198,6 +207,22 @@ PHP_MINIT_FUNCTION(shadow)
+			&& strcmp(data->name, "apc") && strcmp(data->name, "mhash") // exceptions
+		) {
+			php_printf("Shadow extension should be the last one loaded. Extension %s is not loaded yet.\n", (char*)data->name);
+			return FAILURE;
```
I wouldn't say it's a failure. If it's possible, it should be just a warning. Given that the extension's API is not configured to be overridden, it's fine if goes after shadow.

BTW, it'd be even more useful, if we could catch the function/method which extension registers or overrides, but it's configured to be overridden by shadow. In this case we won't have to maintain the list of exceptions above (I'm sure it's far from being complete).

morozov commented on README.md:
```diff
@@ -59,6 +60,9 @@ php.ini parameters for shadow. Default is fine for most cases.
 | shadow.mkdir\_mask | 0755    | Mask used when creating new directories on instances |
 | shadow.debug       | 0       | Debug level (bitmask)                                |
 | shadow.cache\_size | 10000   | Shadow cache size (in bytes, per process)            |
+| shadow.override    | See bundled shadow.ini | Internal functions overridden by shadow |
```
This is now inconsistent with the renamed shadow.ini → zz-shadow.ini. Not sure, however, if we need to enforce it this way.

Commit:
BR-4382: (SI 76716) Diagnostics not writing custom_directory.zip file in Shadow

Update readme and removed default shadow.ini. Users should enable shadow manually and make sure it's loaded in correct order.
